### ISSUE TITLE
get the request information in brownser mode

### ISSUE THREAD

**Issue by Rbsn-joses** [grafana/xk6-browser#1081]
I would like to know if there is any way to include which requests were made in the test process, I know that status codes are generated for requests made during the test, is there a way to get more information such as endpoints? like we use inspect on google?

**Comment by ankur22** [comment]
Hi @Rbsn-joses and @saudixit,

I answered a similar question [here](https://github.com/grafana/xk6-browser/issues/1079), although this feels slightly different.

It sounds like you require access to all http requests during the test runtime, and be able to action on these http requests. So you're not interested in the measurement values at the end of the test run (e.g. request duration). Is this correct?

What are you hoping to solve by being able to wait on all http requests? When do you expect the event to fire?

**Comment by Rbsn-joses** [comment]
Hello, in this case I wanted k6 to send as metrics, which endpoints or calls were made and which responses they returned, so that, when there is an error status code, it alerts which endpoint also gave the error, it would make it easier when it comes to deal with problems in production, because the script would be to simulate user navigation to test whether all the application's functionalities are responding correctly.
An example, use k6 on the login screen and log in as a test, then the metrics are sent to Grafana, method done, http return and which call was made

**Comment by Lerdson** [comment]
@ankur22 

Your comment snippet deals with the high level statistics of web traffic but I believe I can justify the low level access requested here.

I'm currently working with a customer on a react project that makes many requests via JS throughout the user journey. To accurately test the load a single user would put on the application, testing on the browser level would be most appropriate, lest we create inaccurate fixed sets of requests open to falling out of sync with the actual application behaviour.

Testing on the browser level is much more resource intensive than on the protocol level. However we can achieve an excellent balance of both worlds by basing our protocol level tests off the back of results from a real browser traversing the application.

If it were possible to generate a request history (maybe a HAR file, maybe not) then we could run an initial browser test to create an accurate user load recording and then follow up with protocol level tests using the data we've just generated. When our application starts querying more/different endpoints we can trivially re-run the browser test and then run many iterations of protocol tests based off that data. No manually intervention from a human creating a new recording, no load tests falling out of sync with actual application demands, and the ability to scale tests much more without the overhead of a real browser doing it.

The way I look at things, we developers are not omniscient when it comes to how users use our creation, so it's always best to go about things the most flexible way possible. Creating a callback system for acting on requests and responses as they occur would be a great way of achieving it. I'm sure I've seen such a feature in the playwright API, so perhaps this request really comes down to finishing off the mimicking of those advanced features within K6 browser.

**Comment by mdcruz** [comment]
Hi @Lerdson!

One of our k6 champions did a talk this year where he demonstrated using Playwright's recorder functionality, converting it to a k6 browser test, as well as using the recorder to generate a HAR file and then using the community tool, [har-to-k6](https://github.com/grafana/har-to-k6) to convert the HAR file into a k6 protocol test. 

You can find the recording here https://grafana.com/about/events/grafanacon/2023/session/automated-performance-modeling-with-nasa-open-mct-grafana-cloud-k6/ (timestamp is 15:20 onwards). Would this be something that can help with your scenario? 

Let us know please if this fits what you are looking for 💜

**Comment by Lerdson** [comment]
Hi @mdcruz

Thanks for that. I hadn't thought of mixing playwright and k6 and that should work nicely. 

What I would say is just for the purpose of issue tracking for k6 is it seems so very close to filling this need by itself. Relying on playwright for e2e would seem to invalidate some of the implied goals of k6 as a hybrid load testing/e2e multitool.

I'm not entirely certain the features I mentioned are totally inline with the OP's intentions so perhaps I should open a separate issue just to track the absence of a very handy feature?
